IllegalStateException in SegmentOutputStreamImpl

**Problem description**
Writers logs the following warning repeatedly.
```
2017-12-20 07:12:20,141 527693 [clientInternal-4] WARN  i.p.c.s.i.SegmentOutputStreamImpl - f0976f52-7627-478b-8c1c-440cf5ac6516 Failed to connect: 
java.util.concurrent.CompletionException: java.lang.IllegalStateException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:961)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.pravega.common.concurrent.Futures.lambda$delayedFuture$17(Futures.java:466)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: null
	at com.google.common.base.Preconditions.checkState(Preconditions.java:429)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl.lambda$null$4(SegmentOutputStreamImpl.java:522)
	at io.pravega.common.concurrent.Futures.lambda$delayedFuture$19(Futures.java:486)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
	... 10 common frames omitted
```

**Problem location**
SegmentoutputStreamImpl.

**Suggestions for an improvement**

Fix the issue.
Logs: 
[readwriteandscalewithfailovertest.log.log.zip](https://github.com/pravega/pravega/files/1575506/readwriteandscalewithfailovertest.log.log.zip)

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java
@@ -84,6 +84,9 @@ class SegmentOutputStreamImpl implements SegmentOutputStream {
      * Internal object that tracks the state of the connection.
      * All mutations of data occur inside of this class. All operations are protected by the lock object.
      * No calls to external classes occur. No network calls occur via any methods in this object.
+     * Note: In a failure scenario SegmentOutputStreamImpl.State#failConnection can be invoked before
+     * SegmentOutputStreamImpl.State#newConnection is invoked as we do not want connection setup and teardown to occur
+     * within the scope of the lock.
      */
     @ToString(of = {"closed", "exception", "eventNumber"})
     private final class State {
@@ -530,6 +533,7 @@ class SegmentOutputStreamImpl implements SegmentOutputStream {
                          try {
                              connection.send(cmd);
                          } catch (ConnectionFailedException e1) {
+                             // This needs to be invoked here because call to failConnection from netty may occur before state.newConnection above.
                              state.failConnection(e1);
                              throw Lombok.sneakyThrow(e1);
                          }

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java
@@ -530,6 +530,7 @@ class SegmentOutputStreamImpl implements SegmentOutputStream {
                          try {
                              connection.send(cmd);
                          } catch (ConnectionFailedException e1) {
+                             state.failConnection(e1);
                              throw Lombok.sneakyThrow(e1);
                          }
                          return connectionSetupFuture.exceptionally(t -> {

--- a/client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java
+++ b/client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java
@@ -53,12 +53,7 @@ import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.Mockito.doAnswer;
-import static org.mockito.Mockito.inOrder;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.verifyNoMoreInteractions;
-import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.*;
 
 
 public class SegmentOutputStreamTest {
@@ -111,6 +106,26 @@ public class SegmentOutputStreamTest {
 
         cf.getProcessor(uri).connectionDropped(); // simulate a connection dropped
         //Ensure setup Append is invoked on the executor.
+        verify(connection).send(new SetupAppend(2, cid, SEGMENT));
+    }
+
+    @Test(timeout = 10000)
+    public void testConnectAndFailedSetupAppend() throws Exception {
+        UUID cid = UUID.randomUUID();
+        PravegaNodeUri uri = new PravegaNodeUri("endpoint", SERVICE_PORT);
+
+        MockConnectionFactoryImpl cf = new MockConnectionFactoryImpl();
+        ScheduledExecutorService executor = mock(ScheduledExecutorService.class);
+        implementAsDirectExecutor(executor); // Ensure task submitted to executor is run inline.
+        cf.setExecutor(executor);
+
+        MockController controller = new MockController(uri.getEndpoint(), uri.getPort(), cf);
+        ClientConnection connection = mock(ClientConnection.class);
+        doThrow(ConnectionFailedException.class).doNothing().when(connection).send(any(SetupAppend.class));
+        cf.provideConnection(uri, connection);
+        SegmentOutputStreamImpl output = new SegmentOutputStreamImpl(SEGMENT, controller, cf, cid, segmentSealedCallback, RETRY_SCHEDULE);
+        output.reconnect();
+        verify(connection).send(new SetupAppend(1, cid, SEGMENT));
         verify(connection).send(new SetupAppend(2, cid, SEGMENT));
     }
 

